### PR TITLE
Add creation of user defined functions

### PR DESCRIPTION
**Pull request by AlexanderTaeschner**
Add functions to create user defined functions.

Addresses #14

### CODE REVIEW

**Comment by AlexanderTaeschner** [comment]
Ready to be reviewed (T4 was not so difficult with the excellent example given in #14 - thanks!!!).

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 1]
```diff
@@ -265,6 +267,352 @@ protected override DbCommand CreateDbCommand()
...
+        {
+            delegate_function_aggregate_step stepFunction = null;
+            delegate_function_aggregate_final finalFunction = null;
+            if (function != null)
+            {
+                stepFunction = (ctx, _, args) =>
+                {
+                    if (ctx.state == null)
+                    {
+                        ctx.state = seed;
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: If the result can be `null`, this may cause problems. One way around it would be to wrap the value in an type so we know if we've initialized it or not.
```C#
if (ctx.state == null)
    ctx.state = new MyHolder<TContext> { Value = seed };

ctx.state.Value = function(ctx.state.Value);
```

Comment (AlexanderTaeschner) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Changed.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 2]
```diff
@@ -297,6 +299,352 @@ public virtual void CreateCollation<T>(string name, T state, Func<T, string, str
...
+            if (type == typeof(DBNull))
+            {
+                return (T)(object)DBNull.Value;
+            }
+            if (type == typeof(object))
+            {
+                return (T)GetArgumentObject(value);
+            }
+
+            return SqliteDataReader.GetValue<T>(
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: I don't like adding this responsibility directly to `SqliteDataReader`. I wonder if we can put all the type coercing into a base class (possibly `SqliteDataRecord` inheriting `DbDataRecord`), and create a separate derived class that provides APIs of a list of `sqlite3_value` objects.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 3]
```diff
@@ -297,6 +299,352 @@ public virtual void CreateCollation<T>(string name, T state, Func<T, string, str
...
+            }
+
+            return array;
+        }
+
+        private void SetFunctionResult(object result, sqlite3_context ctx)
+        {
+            var type = result.GetType().UnwrapNullableType().UnwrapEnumType();
+
+            SqliteParameter.CoerceValue(
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Similarly, I don't like overloading `SqliteParameter`. We should also look into an abstraction over the concept of a `sqlite3_result` that shares a common base type with `SqliteParameter`.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs (hunk 2 above)
Comment (AlexanderTaeschner) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: I tried the use of a class inheriting from DbDataRecord, but a large number of functions which have to be implemented for the IDataRecord interface, depend on the state of the DataReader, e.g.
```[csharp]
        public override bool IsDBNull(int ordinal)
            => _closed
                ? throw new InvalidOperationException(Resources.DataReaderClosed(nameof(IsDBNull)))
                : !_stepped || _done
                    ? throw new InvalidOperationException(Resources.NoData)
                    : GetSqliteType(ordinal) == raw.SQLITE_NULL;
```
I therefore created a new static class with all the value conversion.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs (hunk 3 above)
Comment (AlexanderTaeschner) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: See above.

Comment (AlexanderTaeschner) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: With the newest version this is refactored into new classes which do the value storage. Since SqliteParameter uses this also for rebinding the value to a new statement/index the base class is a bit more complex than needed for the result storage.

**Comment by bricelam** [comment]
@AlexanderTaeschner any chance you could rebase and squash this on top of `dev`?

**Comment by AlexanderTaeschner** [comment]
Okay, rebased and squashed.

**Comment by bricelam** [comment]
Sorry I'm taking forever on this one. I'm still not completely happy with the factoring. I'm going to play around with the code a bit and see if I can refine the parts I don't like a bit more. Forgive me; I'm an especially picky maintainer on this, [my weekend project](http://www.bricelam.net/2015/04/29/sqlite-on-corefx.html).

**Comment by bricelam** [comment]
Just tested it, and throwing an exception "works", but it feels very incorrect since `sqlite3_step()` never returns. I'll update it to catch any exceptions and call `sqlite3_result_error()`.

**Comment by AlexanderTaeschner** [comment]
Looking at issue #354 the question arrises if we should discard this PR and replace all Create... functions with Add... functions that work independent of the state of the connection (so that they can be issued even when the connection is not yet open).

**Comment by AlexanderTaeschner** [comment]
As addition to Add.. functions maybe also add Remove... ones to delete functions.

**Comment by bricelam** [comment]
#354 is going to require A LOT of design thought.

For now, I'd like to keep the programming model as close to the native APIs as possible. E.g. The connection needs to be open for most operations and they don't persist after the connection is closed.

**Comment by ericsink** [comment]
Oops.  Sorry I missed a mention above.  I keep thinking I should pay more attention to this thread, because it looks interesting.  Anyway, it looks like you moved past that particular question you had for me.

**Comment by bricelam** [comment]
Maybe in 3.0, we implement #354 not by adding new APIs, but by making the existing APIs persist. Whatever we do, I don't expect to tackle #354 any time soon.

**Comment by bricelam** [comment]
@ericsink No worries. I was just checking to see if you've thought about it before. After testing I decided the safest thing to do was catch any exceptions and call `sqlite3_result_error()`.

**Comment by bricelam** [comment]
There we go. There's noting a few more abstractions and levels of indirection can't solve. 😉

**Comment by bricelam** [comment]
I need to add a few more tests around `null` (and `DBNull`) values, but the bulk of it is ready for review.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 4]
```diff
@@ -377,5 +377,164 @@ public virtual void EnableExtensions(bool enable = true)
...
+            {
+                throw new InvalidOperationException(Resources.CallRequiresOpenConnection(nameof(CreateFunction)));
+            }
+
+            delegate_function_scalar func = null;
+            if (function != null)
+            {
+                func = (ctx, user_data, args) =>
+                {
+                    // TODO: Avoid allocation when niladic
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: The TODOs left in this PR are theoretical perf improvements. I didn't want to make them without measuring the improvement since they would complicate the code base.

File: src/Microsoft.Data.Sqlite.Core/SqliteParameter.cs [hunk 5]
```diff
@@ -208,137 +204,11 @@ internal bool Bind(sqlite3_stmt stmt)
...
-                else
-                {
-                    throw new InvalidOperationException(Resources.UnknownDataType(type));
-                }
-
-                _bindActionValid = true;
-            }
-
-            _bindAction(stmt, index);
+            new SqliteParameterBinder(stmt, index, _value).Bind();
```
Comment (AlexanderTaeschner) [inline src/Microsoft.Data.Sqlite.Core/SqliteParameter.cs]: Why don't you store the binding action anymore? This will slow down the reuse of prepared statements,

Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteParameter.cs]: Over-optimization. We haven't done any performance profiling on the library, so we don't actually know what difference it makes. It would be nice to do an actual round of performance tuning this release.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs [hunk 6]
```diff
@@ -387,6 +388,397 @@ public void CreateCollation_works_with_state()
...
+        public void CreateFunction_works_when_parameter_null_but_type_not_nullable()
+        {
+            using (var connection = new SqliteConnection("Data Source=:memory:"))
+            {
+                connection.Open();
+                connection.CreateFunction("test", (long x) => x);
+
+                var result = connection.ExecuteScalar<long>("SELECT test(NULL);");
+
+                Assert.Equal(0L, result);
```
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: This behavior is odd (when specifying NULL for non-nullable parameters), but it matches how SQLite will coerce NULL when you ask for a value of a specific type...

Comment (bricelam) [reply test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: The other option was to throw. I'll discuss with the team.

File: test/Microsoft.Data.Sqlite.Tests/SqliteDataReaderTest.cs [hunk 7]
```diff
@@ -293,6 +305,23 @@ public void GetFieldValue_of_DBNull_works()
...
+            using (var connection = new SqliteConnection("Data Source=:memory:"))
+            {
+                connection.Open();
+
+                using (var reader = connection.ExecuteReader("SELECT 1;"))
+                {
+                    var hasData = reader.Read();
+
+                    Assert.True(hasData);
+                    Assert.Throws<InvalidCastException>(() => reader.GetFieldValue<DBNull>(0));
```
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteDataReaderTest.cs]: In some extremely obscure case, one might be consider a breaking change.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.cs [hunk 8]
```diff
@@ -0,0 +1,897 @@
...
+        /// <typeparam name="T10">The type of the tenth parameter of the function.</typeparam>
+        /// <typeparam name="T11">The type of the eleventh parameter of the function.</typeparam>
+        /// <typeparam name="T12">The type of the twelfth parameter of the function.</typeparam>
+        /// <typeparam name="T13">The type of the thirteenth parameter of the function.</typeparam>
+        /// <typeparam name="T14">The type of the fourteenth parameter of the function.</typeparam>
+        /// <typeparam name="T15">The type of the fifteenth parameter of the function.</typeparam>
+        /// <typeparam name="TAccumulate">The type of the accumulator value.</typeparam>
+        /// <param name="name">The name of the SQL function.</param>
+        /// <param name="func">An accumulator function to be invoked on each element.  Pass null to delete a function.</param>
+        public virtual void CreateAggregate<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, TAccumulate>(string name, Func<TAccumulate, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, TAccumulate> func)
```
Comment (natemcmaster) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.cs]: Good grief that's a lot of type params. I hope this is never used.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.tt [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Data.Sqlite
+{
+    partial class SqliteConnection
+    {
+<#
+    var ordinal = new[] { "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth", "thirteenth", "fourteenth", "fifteenth" };
```
Comment (natemcmaster) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.tt]: Do we really need 15? `Action<T>` and `Func<T>` top out at 8.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs [hunk 10]
```diff
@@ -153,6 +150,7 @@ public override bool NextResult()
             raw.sqlite3_reset(_stmt);
 
             (_stmt, _hasRows) = _stmtQueue.Dequeue();
+            _record = new SqliteDataRecord(_stmt);
```
Comment (natemcmaster) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: Have you profiled the perf impact of this? My concern is that this is the read hot path, and this adds an allocation per row, a potentially huge increase in memory traffic. Unless I missed it, SqliteDataRecord is only used from SqliteDataReader, which means it could stay inside this type and doesn't need to be factored out into a separate type.

Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: We've never profiled this product. We should.

This is only one allocation per result set (typically one per reader), not row. It shouldn't be significant.

No multiple inheritance; composition was the only option.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.tt (hunk 9 above)
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.tt]: lol, I just used the max `Func` would allow on .NET Standard 2.0.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs (hunk 10 above)
Comment (natemcmaster) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: > This is only one allocation per result set

Oh right. NextResult() means next result of a statement, not next row. Ok, my concern is gone.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.tt (hunk 9 above)
Comment (natemcmaster) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.CreateAggregate.tt]: Well, I guess it doesn't hurt.

File: test/Microsoft.Data.Sqlite.Tests/SqliteDataReaderTest.cs (hunk 7 above)
Comment (natemcmaster) [reply test/Microsoft.Data.Sqlite.Tests/SqliteDataReaderTest.cs]: What's the case?

Comment (bricelam) [reply test/Microsoft.Data.Sqlite.Tests/SqliteDataReaderTest.cs]: `reader.GetFieldValue<DBNull>(0);` when the field is not `NULL`. Although I'd like to meet the person calling this with `DBNull`.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs (hunk 6 above)
Comment (bricelam) [reply test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: We decided to throw. I'll make the change.
